### PR TITLE
Final Project Blog: Loop Unrolling in Petr4

### LINKED ISSUE #228: Project Proposal: Loop Unrolliing in the Networking Domain
Samwise Parkinson
Github username: stp59

# Packet Header Parser Unrolling

## Motivation

P4 is a DSL used to program the forwarding plane of networking devices, and it includes a sublanguage for parsing packet headers which resembles a state machine. The sub-language consists of states, which a blockes of code followed by transition statements. Because of the nature of the domain, the target machine languages for these programs is typically not turing complete, whereas it turns out the the parser sub-language [is](https://github.com/p4lang/p4-spec/issues/46). The language specification "resolves" this issue with a vague suggestion that parser loops should be unrolled at compile-time, or that loops that are not clearly finite should be rejected by some sort of static analysis pass. This makes the formal specification and verification of the parser sub-language a challenge.

## Proposal

I wish to understand loop unrolling in this domain more closely. To this end, I will modify bril to resemble selected features of the P4 parser sub-language and implement a version of loop unrolling. The implementation should:

* check the program for certain basic conditions indicating that any loops are finite, and fail with  a helpful error message if not

* compute the number of unrollings needed to make the loops finite

* perform a loop unrolling transformation, and output an equivalent bril program without any loops in the CFG

I will start with very restrictive conditions checking if the loops can be unrolled, and I will try to keep the modified version of bril as simple as possible in order to focus on implementing loop unrolling for a arbitrary number of loop copies. This done, I will search for ways to weaken the conditions checking if the loop is finite, keeping common use-cases of P4's parser language in mind.

## Evaluation

Over the course of implementation, I will compile a collection of programs in the modified version of bril. For positive test cases, I will check that the input and output programs behave the same using the bril interpreter (I may need to implement my own interpreter for the modified version). I will also search for tricky edge cases which test that my conditions for unrolling are implemented correctly. Some of the positive test cases can be derived from real-world P4 programs.

### PR DESCRIPTION
**Pull request by stp59**
closes #228

### CODE REVIEW

File: content/blog/2020-12-18-stp59-final.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+However, loops in the domain of network programming languages have entirely unique significance. Typical hardware targets (e.g. routers) are quite different from traditional CPUs, sometimes requiring statically allocated, straight-line code. As a result, many languages in the networking domain are intended to be terminating, or at least not to be Turing Complete. Loops, then, usually ought to be not much more than syntactic sugar for essentially copy-pasting the same code some `n` times with a small change at each copy. For this reason, loop-unrolling is less an optimization than a fundamental feature of a language's static and dynamic semantics. It is also needed less for the sake of effeciently repeating the same comuptation thousands of times, and more for finding an acceptable version of the source program which can be loaded onto networking hardware.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Always use a comma after "e.g." https://capra.cs.cornell.edu/styleguide/#egie

File: content/blog/2020-12-18-stp59-final.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+This project explores this unique treatment of loop unrolling in the network programming language P4. P4 is used to program network switches and routers, and is of the most practical and widely used languages in the domain. Most of the language resembles basic C without `for` and `while` statements. Loops in P4 come in the form of the parser sub-language, which resembles a finite state machine (FSM) and is used by a network device to extract metadata from in-going packets into useful data structures.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Stray "of" in "is of the most practical"?

File: content/blog/2020-12-18-stp59-final.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+Unfortunately, P4's only [specification](https://p4.org/p4-spec/docs/P4-16-v1.2.0.html) comes in the form of a ~200 page pdf document consisting of vague English descriptions, pseudocode, and examples. For instance, the document acknowledges that targets typically require straight-line code, yet it turns out that the parser sublanguage (the only part that has loops) is [Turing Complete](https://github.com/p4lang/p4-spec/issues/46). All the spec does to address this is make a vague suggestion that this be handled on a per-target basis:
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Capitalize "PDF."

File: content/blog/2020-12-18-stp59-final.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+```
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Use a Markdown blockquote instead of a code block. https://commonmark.org/help/tutorial/05-blockquotes.html

File: content/blog/2020-12-18-stp59-final.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+Although parsers may contain loops, provided some header is extracted on each cycle, the packet itself provides a bound on the total execution of the parser. In other words, under these assumptions, the computational complexity of a P4 program is linear in the total size of all headers, and never depends on the size of the state accumulated while processing data (e.g., the number of flows, or the total number of packets processed).
+
+... 
+
+Architectures may impose (static or dynamic) constraints on the number of parser states that can be traversed for processing each packet. For example, a compiler for a specific target may reject parsers containing loops that cannot be unrolled at compilation time or that may contain cycles that do not advance the cursor.
+```
+
+It is almost unbelievable that the above two blocks comprise the entirety of the P4 spec's treatment of this problem, especially in a turing complete language which targets hardware that is often not turing complete. It is clear that this needs further study.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: In the programming language English, "blocks" are called "paragraphs." 😃

Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Capitalize "Turing."

File: content/blog/2020-12-18-stp59-final.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+Petr4 is a long-term project which strives to build a formally verified end-to-end compiler for P4, along with [formal operational semantcs](https://popl21.sigplan.org/details/POPL-2021-research-papers/41/Petr4-Formal-Foundations-for-P4-Data-Planes) for the surface syntax. As it exists now, it consists of a [front-end](https://github.com/cornell-netlab/petr4) with a type-checker and evaluator in OCaml, from which formal semantics have been extracted on pencil-and-paper.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Since this is the first mention of Petr4, make the term itself a link.

File: content/blog/2020-12-18-stp59-final.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+## Implementation
+
+### Static Semantics of Parser Loops
+
+The surface syntax of P4 parsers consists of local declarations followed by states. These states are comprised of a highly restricted set of statements (variable assignments and method calls only) followed by a transition statement, indicating a jump to another state. Transitions may be conditional, branching to arbitrarily many different states. The result is a system in which arbitrary irreducible control flow arises quite easily.
+
+However, the common case is to use a natural loop in the parser to populate various data structures provided by the language. In particular, parser loops that extract from the packet on each iteration are guaranteed to terminate due to the fact that the packet has a finite length. With these considerations, I am able to provide a concrete proposal for a static analysis tool of parser loops. Accepted parsers should meet the following conditions:
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: A code example could _really_ help here to illustrate what these "natural loops" look like. Is there a short one you can include?

File: content/blog/2020-12-18-stp59-final.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+1) They should have reducible control flow
+2) Every distinct natural loop should include a block which both extracts from the packet and dominates all paths to the loop header
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: I'm not sure what it means to "dominate" a path. We usually say that CFG nodes dominate other CFG nodes. Maybe you mean "dominates the loop header"?

File: content/blog/2020-12-18-stp59-final.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+My implementation runs these checks, providing the user with the options to fail when they are not satisfied or merely to emit a warning and produce unmodified code. Surprisingly, this turned out to be the most computationally heavy part of the implementation, not the actual transformation. While natural loops are easily identified about a backedge in the control flow, arbitrary loops require a full traversal of the control flow graph (CFG). I used [Tarjan's Strongly Connected Components Algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm), which boasts linear runtime, to compute strongly-connected components of an arbitrary CFG. I then check that they all satisfied the criteria for forming a natural loop: that there exists a unique header node `h` which dominates all other nodes in the strongly-connected component and which is the only node with incoming edges from outside of the loop. Having identified the natural loops, I then traverse them to ensure that every path back to the header consumes the packet. Again, this ensures termination as packets are guaranteed to be finite.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md:195]: Again, examples might be nice here, especially of programs that your checker would warn/error on.

File: content/blog/2020-12-18-stp59-final.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+For correctness, I begin with a minimal collection of toy p4 programs which I developed for the sake of test-driven development of loop unrolling. Residing in `petr4/unroll-test`, these are edge-case control flow programs which are primarily expected-fail for the static analysis. In addition to these, Petr4 conveniently includes a robust test harness which runs the front-end on the entirety of p4c's benchmark suite. Combined with some custom test cases from the front-end development, this runs over 100 realistic p4 programs, each on several realistc packets, all in under a few seconds. In order to acquire reasonable evidence of the correctness of my transformation, I merely had to insert a call to my unroller in this test harness before type checking. After some work, my tool achieves the same correctness as the Petr4's evaluater on unmodified code. The only limitation of this approach is that it does not include any negative tests, so I am much more sure of the correctness of my code transformation tool than I am of my static analysis tool.
```
Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: Link to the directory of tests on GitHub?

Comment (sampsyo) [inline content/blog/2020-12-18-stp59-final.md]: I again think that including some small code examples here could really help illustrate what's going on.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Great! This is a cool outcome, and I'm glad you were able to make it work within the real Petr4 codebase. :tada: I have a few suggestions within—most notably about including some P4 code examples to illustrate what a loop looks like in the parsers you're processing.

**Comment by stp59** [comment]
I believe I have addressed all of your suggested changes!

**Review by sampsyo — APPROVED** [review]
Looks good; thanks!!
